lib/streamlit/web/server/oauth_authlib_routes.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import json
+from typing import Any
+from urllib.parse import urlparse
+
+import tornado.web
+
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.secrets import secrets_singleton
+from streamlit.user_info import decode_provider_token
+from streamlit.web.server.oidc_mixin import TornadoOAuth, TornadoOAuth2App
+from streamlit.web.server.server_util import AUTH_COOKIE_NAME
+
+
+class AuthCache:
+    def __init__(self):
+        self.cache = {}
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def set(self, key, value, expires_in):
+        self.cache[key] = value
+
+    def get_dict(self):
+        return self.cache
+
+    def delete(self, key):
+        self.cache.pop(key, None)
+
+
+auth_cache = AuthCache()
+
+
+def create_oauth_client(provider: str) -> tuple[TornadoOAuth2App, str]:
+    if secrets_singleton.load_if_toml_exists():
+        auth_section = secrets_singleton.get("auth")
+        if auth_section:
+            redirect_uri = auth_section.get("redirect_uri", None)
+            config = dict(auth_section.to_dict())
+        else:
+            config = {}
+            redirect_uri = "/"
+    else:
+        config = {}
+        redirect_uri = "/"
+
+    provider_section = config.setdefault(provider, {})
+    provider_client_kwargs = provider_section.setdefault("client_kwargs", {})
+    if "scope" not in provider_client_kwargs:
+        provider_client_kwargs["scope"] = "openid email profile"
+    if "prompt" not in provider_client_kwargs:
+        provider_client_kwargs["prompt"] = "select_account"
+
```

Comment:
for me, using Microsoft AD, I need to also have this row:
```
provider_client_kwargs["code_challenge_method"] = "S256"
```
to enable the generation of the PKCE code.
Can this be added as parameter in the next relase?

Reply:
Hey @mauro-mascia, thank you for your feedback!

We already set "code_challenge_method" to "S256": if it is supported by OIDC provider, we look to `code_challenge_methods_supported` value for `.well-known` configuration. 
So most probably it is already work. 

https://github.com/streamlit/streamlit/pull/8786/files#diff-e7111ded76c5b321c2808266062e60890d8bca6d1b202533ef2e6ec62befd5a8R33-R37

Also you can always set client_kwargs explicitly in secrets.toml auth section of concrete provider:
```toml
[auth]
redirect_uri = "http://localhost:8501/oauth2callback"
cookie_secret = "your_cookie_secret_here"

[auth.auth0]
client_id = "AAA"
client_secret = "BBB"
server_metadata_url = "https://cccc.us.auth0.com/.well-known/openid-configuration"
client_kwargs = { 'prompt' = 'consent' }
```

In the next weeks we will work to make it more explicit!

Reply:
My last issue is that when I return from the callback I correctly got the `code=1.ASAAGgwVWwGhb...` (I'm on Microsoft AD) but it does not store any cookie. I think that `AuthCallbackHandler` should be called but it does not go here. Any idea?

Reply:
Is there any chance you are behind the proxy? 
You should configure `redirect_uri` in the `secrets.toml`

Something like: 
```toml
[auth]
redirect_uri = "http://localhost:8501/oauth2callback"
cookie_secret = "your_cookie_secret_here"

[auth.microsoft]
client_id = "blablabla"
```
and the `redirect_uri` should be the same as that configured in your Microsoft AD app. 

Also, does Microsoft AD support OIDC protocol?
Because this feature is intended to support only OIDC, I tested it with `Microsoft Entra ID`, and it works with it.

Reply:
Thanks @kajarenc this question was related with another comment https://github.com/streamlit/streamlit/pull/8786#discussion_r1838506244
and the fact that the redirect has been configured to http://localhost:8501/callback. For that reason it wasn't working. Maybe we need to specify that this is the required return URL somewhere if not already specified (I didn't see it until watching the code). Anyway it's ok for me to set it to /oauth2callback thanks.